### ISSUE TITLE
RedirectService

### ISSUE THREAD

**Issue by trustin** [line/armeria#774]
Related: https://github.com/line/centraldogma/pull/33

It would be useful when a user wants to send a redirect response on a certain request:

```java
ServerBuilder sb = ...;
sb.service("/old-location", new RedirectService("/new-location"));
sb.service("/old-location/{name}", new RedirectService(ctx -> "/new-location/" + ctx.pathParam("name")));
sb.service("/old-location/{name}", new RedirectService("/new-location/{name}"));
sb.service("/old-location", new RedirectService(RedirectType.PERMANENT, "/new-location"));
```

/cc @hyangtack

**Comment by trustin** [comment]
@abdulazizali77 Yes, of course. Thanks for your interest!

> should we only implement doGet and doPost? Im completely unsure whether OPTIONS, PUT, HEAD, DELETE should have a redirect, please advise

We could probably allow a user specify the methods to redirect? By default we could redirect for all methods.

> should we only support 301, 307, 308?

I think so. Please feel free to add support for more cases though if there are any.

> how should we handle the case /new-path that a RedirectService points to doesnt have a configured Service before we call build?

That's a good question. Ideally, we could add a listener to the server in `serviceAdded()` to get notified when the server if fully built and check if there's a matching path mapping actually for validation purpose. However, I think it's just fine to allow a user to specify the target path which leads to 404, because he or she might have done it intentionally.

> am i correct to believe that this will have no impact on GrpcServiceBuilder or THttpService . I am unsure if grpc or thrift has mechanism for redirect of its own.

I think `GrpcServiceBuilder` or `THttpService` is not relevant.

**Comment by trustin** [comment]
Fixed via #945
